pkg/cmd/server/origin/audit.go

```diff
@@ -0,0 +1,202 @@
+	"k8s.io/apiserver/pkg/audit/policy"
+	"k8s.io/apiserver/pkg/endpoints/filters"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithAuthFallbackAudit decorates a http.Handler with a fallback audit, logging
+// information only when the original one did was not triggered.
+// This needs to be used with WithAuditTriggeredMarker, which wraps the original
+// audit filter.
+func WithAuthFallbackAudit(handler http.Handler, requestContextMapper request.RequestContextMapper,
+	sink audit.Sink, policy policy.Checker) http.Handler {
+	if sink == nil || policy == nil {
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		respWriter := decorateResponseWriter(w, getUsername(req), req, requestContextMapper, sink, policy)
+		handler.ServeHTTP(respWriter, req)
+	})
+}
+
+// decorateResponseWriter is a copy method from the upstream audit, adapted
+// to work with the fallback audit mechanism.
+func decorateResponseWriter(responseWriter http.ResponseWriter, username string, req *http.Request,
+	requestContextMapper request.RequestContextMapper, sink audit.Sink, policy policy.Checker) http.ResponseWriter {
+	delegate := &auditResponseWriter{
+		ResponseWriter:       responseWriter,
+		username:             username,
+		request:              req,
+		requestContextMapper: requestContextMapper,
+		sink:                 sink,
+		policy:               policy,
+	}
+	// check if the ResponseWriter we're wrapping is the fancy one we need
+	// or if the basic is sufficient
+	_, cn := responseWriter.(http.CloseNotifier)
+	_, fl := responseWriter.(http.Flusher)
+	_, hj := responseWriter.(http.Hijacker)
+	if cn && fl && hj {
+		return &fancyResponseWriterDelegator{delegate}
+	}
+	return delegate
+}
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+
+// auditResponseWriter intercepts WriteHeader, sets it in the event.
+type auditResponseWriter struct {
+	http.ResponseWriter
+	once                 sync.Once
+	username             string
+	request              *http.Request
+	requestContextMapper request.RequestContextMapper
+	sink                 audit.Sink
+	policy               policy.Checker
+}
+
+func (a *auditResponseWriter) Write(bs []byte) (int, error) {
+	a.processCode(http.StatusOK) // the Go library calls WriteHeader internally if no code was written yet. But this will go unnoticed for us
+	return a.ResponseWriter.Write(bs)
+}
+
+func (a *auditResponseWriter) WriteHeader(code int) {
+	a.processCode(code)
+	a.ResponseWriter.WriteHeader(code)
+}
+
+func (a *auditResponseWriter) processCode(code int) {
+	a.once.Do(func() {
+		ctx, ok := a.requestContextMapper.Get(a.request)
+		if !ok {
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("no context found for request"))
+			return
+		}
+
+		// if there already exists an audit event in the context we don't need to do anything
+		if ae := request.AuditEventFrom(ctx); ae != nil {
```

Comment:
Yeah, me neither. I was thinking about making this whole thing a helper function in audit upstream, wdyt?